### PR TITLE
express-notes-api | ChiaYu Lin

### PR DESCRIPTION
**Pull request by cyulin17**
express-notes-api-get-1
express-notes-api-get-2
express-notes-api-get-3
express-notes-api-get-4

express-notes-api-post-1
express-notes-api-post-2
express-notes-api-post-3

express-notes-api-delete-1
express-notes-api-delete-2
express-notes-api-delete-3
express-notes-api-delete-4

express-notes-api-put-1
express-notes-api-put-2
express-notes-api-put-3
express-notes-api-put-4
express-notes-api-put-5

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+app.get('/api/notes', function (req, res) {
+
+  const array = [];
+
+  for (const prop in jsonObj.notes) {
+    array.push(jsonObj.notes[prop]);
+  }
+  res.json(array);
+});
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:14]: "Clients can GET a list of notes" looks good.

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  if (Number.isInteger(Number(idNum)) === true && Number(idNum) > 0 && jsonObj.notes[idNum] !== undefined) {
+    const result = jsonObj.notes[idNum];
+    return res.status(200).send(result);
+  } else if (Number.isInteger(Number(idNum)) === true && Number(idNum) > 0) {
+    return res.status(404).json({ error: 'cannot find note with id ' + idNum });
+  } else {
+    return res.status(400).json({ error: 'id must be a positive integer' });
+  }
+
+});
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:29]: "Clients can GET a single note by id" looks good.

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    if (err) {
+      res.status(500).json({ error: 'An unexpected error occurred.' });
+    } else {
+      res.status(201).send(newPost);
+    }
+
+  });
+
+});
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:55]: "Clients can POST a new note" looks good.

File: express-notes-api/index.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+  var deleteData = JSON.stringify(jsonObj, null, 2);
+  fs.writeFile('data.json', deleteData, { encoding: 'utf8' }, err => {
+    if (err) {
+      res.status(500).json({ error: 'An unexpected error occurred.' });
+    } else {
+      res.sendStatus(204);
+    }
+  });
+});
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:77]: "Clients can DELETE a note by id" look good.

File: express-notes-api/index.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  var updateData = JSON.stringify(jsonObj, null, 2);
+  fs.writeFile('data.json', updateData, { encoding: 'utf8' }, err => {
+    if (err) {
+      res.status(500).json({ error: 'An unexpected error occurred.' });
+    } else {
+      res.status(200).send(jsonObj.notes[updateId]);
+    }
+  });
+
+});
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:103]: "Clients can replace a note (PUT) by id" looks good.

**Review by TimelikeClosure — APPROVED** [review]
Excellently done!

**Comment by cyulin17** [comment]
> Excellently done!

Thank you!

**Pull request merged**